### DISCUSSION TITLE
Phase 41.4 — InputSanitizer: Feature Squeezing & Denoising Defense Deep-Dive

### DISCUSSION THREAD

**Discussion by web3guru888** [web3guru888/asi-build#836] [Show and tell]
# Phase 41.4 — InputSanitizer: Implementation Deep-Dive

## Model-Agnostic Defense Layer

The InputSanitizer provides defense at the input level — transforming or rejecting inputs before they reach the model. This is the most practical first line of defense for deployed systems, requiring no model retraining.

## Feature Squeezing Detection (Xu et al., 2018)

The key insight: adversarial perturbations are fragile and often destroyed by simple input transformations. If the model's prediction changes significantly after squeezing, the input is likely adversarial.

**Detection pipeline:**
```
1. Compute model prediction on original input: p = f(x)
2. For each squeezer s_i (bit-depth, smoothing, JPEG):
   a. Compute squeezed prediction: p_i = f(s_i(x))
   b. Compute L1 distance: d_i = ‖p - p_i‖₁
3. If max(d_i) > threshold → flag as adversarial
```

**Three squeezing operations:**
- **Bit-depth reduction** (e.g., 8-bit → 4-bit): Quantizes pixel values, destroying small perturbations
- **Spatial smoothing** (median filter 3×3): Removes high-frequency adversarial patterns
- **JPEG compression** (quality 75): Lossy compression removes subtle perturbations

## Denoising Autoencoders as Defenses

Train an autoencoder to reconstruct clean images from potentially adversarial inputs:

```python
class DenoisingDefense:
    def __init__(self, autoencoder, threshold):
        self.ae = autoencoder  # trained on clean data
        self.threshold = threshold
    
    def sanitize(self, x):
        x_reconstructed = self.ae(x)
        mse = torch.mean((x - x_reconstructed) ** 2)
        
        if mse > self.threshold:
            # High reconstruction error → likely adversarial
            return x_reconstructed, True  # return denoised version
        return x, False  # clean, pass through
```

## Randomized Defenses

Adding randomness breaks gradient-based attacks:
- **Random resizing**: Resize input to random dimensions before padding back
- **Random padding**: Add random zero-padding to input boundaries
- **Pixel deflection**: Randomly redistribute pixel values in local neighborhoods

**Caveat:** Athalye et al. (2018) showed that many randomized defenses can be broken by Expectation over Transformation (EOT). Our implementation includes EOT-aware evaluation.

## The Adaptive Attack Problem

Tramèr et al. (2020) demonstrated that many input-level defenses create a false sense of security because they are evaluated only against non-adaptive attacks. Our implementation addresses this by:

1. **BPDA (Backward Pass Differentiable Approximation)**: For non-differentiable transformations, approximate the gradient
2. **EOT (Expectation over Transformation)**: Average gradients over multiple randomized defense applications
3. **Adaptive threshold search**: Binary search for minimum perturbation that evades detection

## Detection vs Sanitization Trade-offs

| Strategy | Pros | Cons |
|----------|------|------|
| **Detect & Reject** | No accuracy loss on clean data | Service availability impact |
| **Detect & Sanitize** | Always produces output | Possible accuracy degradation |
| **Always Sanitize** | Simplest deployment | Clean accuracy loss |
| **Ensemble Detection** | Best detection rate | Computational overhead |

Our default recommendation: **Detect → Sanitize if suspicious → Reject if high confidence adversarial**

---
*See issue #828 for full specification. Part of [Phase 41 — Adversarial Robustness & Security Intelligence](../discussions/824).*
